packages/back-end/src/models/SegmentModel.ts

```diff
@@ -47,4 +51,49 @@ export class SegmentModel extends BaseClass {
   ): Promise<SegmentInterface[]> {
     return await this._find({ datasource: datasourceId });
   }
+
+  public async getByFactTableId(
+    factTableId: string
+  ): Promise<SegmentInterface[]> {
+    return await this._find({ factTableId });
+  }
+
+  protected migrate(legacySegment: LegacySegmentInterface): SegmentInterface {
+    // if legacySegment doesn't have a type, it's a legacy, which only allowed SQL
+    return { ...legacySegment, type: legacySegment.type || "SQL" };
+  }
+
+  protected async customValidation(segment: SegmentInterface): Promise<void> {
+    if (segment.type === "SQL") {
+      if (!segment.sql) {
+        throw new Error(
+          `${segment.name} is a SQL type Segment, but contains no SQL value`
+        );
+      }
+      if (segment.factTableId) {
```

Comment:
This should not be an error. A user might switch the type of a segment without clearing out previous fields. Also remove the check for filters below this and the check for SQL in the type=fact branch